### PR TITLE
Multi-tensor LAMB

### PR DESCRIPTION
**Pull request by MoisesHer**
Multi-tensor LAMB Optimizer  **(in development / debugging)**

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete
- [X] All changes have test coverage:
- A test of the multi-LAMB optimizer is added in tests/python/unittest/test_optimizer:test_multilamb
- [X] Code is well-documented: 
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [X] Added Mxnet operator (CPU & GPU): _multi_lamb_update (and mix precision version _multi_mp_lamb_update). Given weights and gradients of multiple tensors, it updates all of them in parallel
- [X] Python LAMB optimizer was updated to launch the multi-tensor version if MXNET_OPTIMIZER_AGGREGATION_SIZE > 1

### CODE REVIEW

**Comment by leezu** [comment]
What's the relation with https://github.com/apache/incubator-mxnet/pull/16715/? #16715 recently added 'LAMB' optimizer to `python/mxnet/optimizer/optimizer.py`. Your code is currently in conflict.

**Comment by MoisesHer** [comment]
> What's the relation with #16715? #16715 recently added 'LAMB' optimizer to `python/mxnet/optimizer/optimizer.py`. Your code is currently in conflict. Please resolve the conflict by merging or rebasing on master.
I merged master, so conflict is resolved. This operator is similar to #16715 but instead of updating a single Tensor, it updates multiple-tensors simultaneously. Thus, it expose more parallelism.

**Comment by leezu** [comment]
Thanks for the clarification. Is it necessary to expose it as a separate `multiLamb` optimizer? Can it be integrated with #16715? Why do we need the less parallel implementation? Or are there other differences? Sorry, I didn't read through the code yet.

File: python/mxnet/ndarray/contrib.py [hunk 1]
```diff
@@ -605,3 +605,27 @@ def multi_mp_adamw_update(weights, grads, mean, var, weights32, rescale_grad, lr
                                                     etas=etas,
                                                     name=name,
                                                     **kwargs)
+
+def multi_lamb_update(weights, grads, mean, var, temp_g, step_count,
+                      out=None, name=None, num_tensors=0, **kwargs):
```
Comment (eric-haibin-lin) [inline python/mxnet/ndarray/contrib.py]: Could you add some documentation on the input types for these two functions? Otherwise it shows nothing on the API website

File: python/mxnet/optimizer/optimizer.py [hunk 2]
```diff
@@ -1051,6 +1052,95 @@ def update_multi_precision(self, index, weight, grad, state):
         self._update_impl(index, weight, grad, state,
                           multi_precision=use_multi_precision)
 
+@register
+class MultiLAMB(Optimizer):
+    """multiLAMB optimizer.
+    """
+    def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: Can we set the default value of lower_bound = None, upperbound = None, and bias_correction=True? Both TF and Pytorch are using the version that does not bound the value of r1 and perform bias correction by default. We should use the same interface as the existing lamb optimizer: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/optimizer/optimizer.py#L1249-L1254

I noticed that the GPU kernel always performs upper/lower bound clipping on r1. Can we add a branch to skip clipping if upper/lower bound = -1?

File: src/operator/contrib/multi_lamb-inl.h [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+      Shape1(kernel_params.ntensors), s);
+    pos_wspace += kernel_params.ntensors * sizeof(float);
+    Tensor<xpu, 1, int> block_to_tensor(reinterpret_cast<int*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.nchunks), s);
+    pos_wspace += kernel_params.nchunks * sizeof(int);
+    Tensor<xpu, 1, int> block_to_chunk(reinterpret_cast<int*>(&workspace[pos_wspace]),
+      Shape1(kernel_params.nchunks), s);
+
+    MultiSumSqRun<xpu>(weights, kernel_params.ntensors, r1.dptr_, s);
+    call_kernel1<MPDType, DType>(s, kernel_params, param, block_to_tensor.dptr_,
```
Comment (eric-haibin-lin) [inline src/operator/contrib/multi_lamb-inl.h]: It looks like MultiSumSq uses atomic add and introduces non-determinism.. We should probably provide a way for users to have deterministic updates. I don't think we need a MultiLAMB optimizer, but rather we reuse the LAMB optimizer class and dispatch to the multi-lamb ops if aggregate size is larger than 1. 
This way, if a user wants deterministic updates, he/she can at least achieve that via setting the aggregation size env var.

File: python/mxnet/optimizer/optimizer.py [hunk 4]
```diff
@@ -1051,6 +1052,95 @@ def update_multi_precision(self, index, weight, grad, state):
...
+        self.upper_bound = upper_bound
+        self.bias_correction = bias_correction
+        self.aggregate_num = max(1, min(50, int(os.getenv('MXNET_OPTIMIZER_AGGREGATION_SIZE', "50"))))
+
+    def create_state(self, index, weight):
+        stype = weight.stype
+        dtype = weight.dtype
+        return (zeros(weight.shape, weight.context, dtype=dtype, stype=stype), # mean
+                zeros(weight.shape, weight.context, dtype=dtype, stype=stype), # variance
+                zeros(weight.shape, weight.context, dtype=dtype, stype=stype)) # temp_g
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: I am a bit confused for the use of temp_g here. Is this mainly used as a temporarly workspace? Is the value of temp_g at step `t-1` used by the update at step `t`? If so would it be possible to allocate g in the op via resource request? Otherwise the state size is increased by 50% when saving the training state. On the other hand, it is inconsistent with the existing LAMB optimizer which has (mean, var) as the state.

File: src/operator/contrib/multi_lamb-inl.h [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    DMLC_DECLARE_FIELD(wd)
+    .set_default(0.0f)
+    .describe("Weight decay augments the objective function with a "
+              "regularization term that penalizes large weights. "
+               "The penalty scales with the square of the magnitude of each weight.");
+    DMLC_DECLARE_FIELD(rescale_grad)
+    .set_default(1.0f)
+    .describe("Gradient rescaling factor");
+    DMLC_DECLARE_FIELD(lower_bound)
+    .set_default(1e-3f)
```
Comment (eric-haibin-lin) [inline src/operator/contrib/multi_lamb-inl.h]: We can set the default value of lower_bound and upper_bound to -1. And in the GPU kernel, we skip the clipping if the value is -1.

File: src/operator/contrib/multi_lamb.cu [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+  MPDType r_mean[ILP_LAMB];
+  MPDType r_var[ILP_LAMB];
+  MPDType r_g[ILP_LAMB];
+
+  for (size_t i=startPos; i < stopPos && i < kernel_params.sizes[tensorID];
+                                                  i+= blockDim.x*ILP_LAMB) {
+#pragma unroll
+      for (int ii = 0; ii < ILP_LAMB; ii++) {
+          int load_pos = i + ii*blockDim.x;
+          if (load_pos < stopPos && load_pos < kernel_params.sizes[tensorID]) {
```
[the remark this answers is not in the document]
Comment (MoisesHer) [reply src/operator/contrib/multi_lamb.cu]: done, thanks

File: python/mxnet/optimizer/optimizer.py (hunk 2 above)
Comment (MoisesHer) [reply python/mxnet/optimizer/optimizer.py]: done, thanks

File: python/mxnet/optimizer/optimizer.py (hunk 4 above)
Comment (MoisesHer) [reply python/mxnet/optimizer/optimizer.py]: done, thanks

File: src/operator/contrib/multi_lamb-inl.h (hunk 5 above)
Comment (MoisesHer) [reply src/operator/contrib/multi_lamb-inl.h]: done, thanks

**Comment by MoisesHer** [comment]
> Ping @MoisesHer Is the MultiLAMB optimizer a generalization of LAMB optimizer? If so, why do we keep the LAMB optimizer? If not, please add documentation or a reference to the docstring. Thank you!

Sorry, since we were having different numbers of states I was not sure how to reuse the previous optimizer. After Haibin suggestion, now we have same numbers of states.
I have merged both optimizers. If aggregation number is <=1 or gradients/weights is a unique NDArray, the single-tensor implementation is used, otherwise, the multi-tensor implementation is used.

**Comment by eric-haibin-lin** [comment]
could you update the code based on https://github.com/apache/incubator-mxnet/pull/17002/ ?

File: python/mxnet/optimizer/optimizer.py [hunk 7]
```diff
@@ -1259,53 +1260,99 @@ def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
...
-
-        if multi_precision:
-            mean, var = state[1]
-            weight32 = state[0]
-            g = mp_lamb_update_phase1(weight, grad, mean, var, weight32, wd=wd, **kwargs)
+        if self.aggregate_num <= 1 or not isinstance(index, (tuple, list)):
+            assert(isinstance(weight, NDArray))
+            assert(isinstance(grad, NDArray))
+            self._update_count(index)
+            lr = self._get_lr(index)
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: let's pass a list of lr, wd to the ops since they can be different.

File: src/operator/contrib/multi_lamb-inl.h [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+};
+
+template<typename DType, typename MPDType>
+struct MultiLAMBKernelParam {
+  static const int N = 45;
+  size_t ntensors;
+  size_t max_size;
+  size_t total_size;
+  size_t sizes[N];
+  size_t tensor2temp_g[N];
```
Comment (eric-haibin-lin) [inline src/operator/contrib/multi_lamb-inl.h]: Can this op handle more than 45 tensors? Is it safe to sage MXNET_OPTIMIZER_AGGREGATION_SIZE > 45?

File: python/mxnet/optimizer/optimizer.py [hunk 9]
```diff
@@ -1259,53 +1260,101 @@ def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
...
                   'rescale_grad': self.rescale_grad}
 
-        if self.clip_gradient:
-            kwargs['clip_gradient'] = self.clip_gradient
-
-        if multi_precision:
-            mean, var = state[1]
-            weight32 = state[0]
-            g = mp_lamb_update_phase1(weight, grad, mean, var, weight32, wd=wd, **kwargs)
+        if self.aggregate_num <= 1 or not isinstance(index, (tuple, list)):
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: MXNET_OPTIMIZER_AGGREGATION_SIZE=1 does not seem to work because weight will be a list instead of NDArray

File: src/operator/contrib/multi_lamb-inl.h [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    .set_default(1)
+    .describe("Number of tensors");
+  }
+};
+
+template<typename ParamType, int input_stride>
+inline bool MultiLAMB_InferShape(const nnvm::NodeAttrs& attrs,
+                                 mxnet::ShapeVector *in_attrs,
+                                 mxnet::ShapeVector *out_attrs) {
+  const ParamType& param = dmlc::get<ParamType>(attrs.parsed);
```
Comment (eric-haibin-lin) [inline src/operator/contrib/multi_lamb-inl.h]: Should we also assert the number of tensors to be fewer than 45?

File: python/mxnet/optimizer/optimizer.py [hunk 11]
```diff
@@ -1259,53 +1260,121 @@ def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-6,
...
+              t = self._index_update_count[index]
+              weight_ptr = weight
+              grad_ptr = grad
+              if multi_precision:
+                mean, var = state[1]
+                weight32 = state[0]
+              else:
+                mean, var = state
+            else:
+              assert(len(index)==self.aggregate_num)
```
Comment (eric-haibin-lin) [inline python/mxnet/optimizer/optimizer.py]: maybe it's cleaner to have one line for this case:
```
if isinstance(index, (tuple, list)):
    index, weight, grad, state = index[0], weight[0], grad[0], state[0]

# then followed by line 1278 - 1290 shared by both cases
```

File: src/operator/contrib/multi_lamb.cc [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+                               kernel_params,
+                               r1, r2,
+                               temp_g,
+                               param.lower_bound, param.upper_bound,
+                               req);
+}
+
+DMLC_REGISTER_PARAMETER(MultiLAMBParam);
+
+std::vector<std::string> LAMBParamToVector(uint32_t num_args, const char *pName[], size_t nParams) {
```
Comment (eric-haibin-lin) [inline src/operator/contrib/multi_lamb.cc]: nit: pName -> p_name

**Comment by eric-haibin-lin** [comment]
Thank you @MoisesHer for addressing all the review comments
